src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+{
+    if handle == INVALID_HANDLE_VALUE {
+        return Ok(INVALID_HANDLE_VALUE);
+    }
+
+    unsafe {
+        let mut new_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let ok = kernel32::DuplicateHandle(*CURRENT_PROCESS_HANDLE as HANDLE, handle,
+                                           other_process, &mut new_handle,
+                                           0, winapi::FALSE, flags);
+        if ok == winapi::FALSE {
+            Err(WinError::last("DuplicateHandle"))
+        } else {
+            Ok(new_handle)
+        }
+    }
+}
+
+// duplicate a handle in the current process
+fn dup_handle(handle: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process(handle, unsafe { kernel32::GetCurrentProcess() })
+}
+
+// duplicate a handle to the target process
+fn dup_handle_to_process(handle: HANDLE, other_process: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process_with_flags(handle, other_process, winapi::DUPLICATE_SAME_ACCESS)
+}
+
+// duplicate a handle to the target process, closing the source handle
+fn move_handle_to_process(handle: HANDLE, other_process: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process_with_flags(handle, other_process, winapi::DUPLICATE_CLOSE_SOURCE | winapi::DUPLICATE_SAME_ACCESS)
+}
+
+// get the current process handle as a handle in another process
+#[allow(dead_code)]
+fn current_process_handle_to_process(other_process: HANDLE) -> Result<HANDLE,WinError> {
+    unsafe {
+        let mut new_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let ok = kernel32::DuplicateHandle(kernel32::GetCurrentProcess(), kernel32::GetCurrentProcess(),
+                                           other_process, &mut new_handle,
+                                           winapi::PROCESS_DUP_HANDLE, winapi::FALSE, 0);
+        if ok == winapi::FALSE {
+            Err(WinError::last("DuplicateHandle"))
+        } else {
+            Ok(new_handle)
+        }
+    }
+}
+
+fn create_overlapped() -> Box<winapi::OVERLAPPED> {
+    Box::new(winapi::OVERLAPPED {
+        Internal: 0,
+        InternalHigh: 0,
+        Offset: 0,
+        OffsetHigh: 0,
+        hEvent: 0 as HANDLE
+    })
+}
+
+fn reset_overlapped(ov: &mut winapi::OVERLAPPED) {
+    ov.Internal = 0;
+    ov.InternalHigh = 0;
+    ov.Offset = 0;
+    ov.OffsetHigh = 0;
+    ov.hEvent = 0 as HANDLE;
+}
+
+
+#[derive(Debug, PartialEq)]
+struct WinHandle {
+    h: HANDLE
+}
+
+unsafe impl Send for WinHandle { }
+unsafe impl Sync for WinHandle { }
+
+impl Drop for WinHandle {
+    fn drop(&mut self) {
+        self.close();
+    }
+}
+
+impl Default for WinHandle {
+    fn default() -> WinHandle {
+        WinHandle { h: INVALID_HANDLE_VALUE }
+    }
+}
+
+impl Deref for WinHandle {
+    type Target = HANDLE;
+
+    #[inline]
+    fn deref(&self) -> &HANDLE {
+        &self.h
+    }
+}
+
+impl WinHandle {
+    fn new(h: HANDLE) -> WinHandle {
+        WinHandle { h: h }
+    }
```

Comment:
> However, this got me wondering: is it indeed unsafe to create a WinHandle from a potentially invalid pointer? We never actually dereference these; but rather just pass them back to the Windows API. I have a growing suspicion that the system might actually be keeping track of valid handle addresses; and erroring out cleanly if we pass an invalid one, rather than trying to actually use it and blowing up in interesting ways... Can you verify this?

Handles are not pointers; they're essentially tagged indices.  Invalid handles are handled and tracked; the Windows API will get you an invalid argument error or similar if an invalid handle is passed in.  But they are definitely not pointers; dereferencing a pointer is not a valid operation.

Honestly -- I don't really want to muck with this pull request much any more, unless there are actual functional (e.g. broken code) reasons to do so.  I'm happy to try to fix things up later or leave to others to improve abstractions or cleanliness, but at this point I really just want to land this, or I'm going to need to move on.